flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetAggFunction.scala

```diff
@@ -68,23 +72,16 @@ class DataSetAggFunction(
   override def reduce(records: Iterable[Row], out: Collector[Row]): Unit = {
 
     // create accumulators
-    var i = 0
-    while (i < aggregates.length) {
-      accumulators(i) = aggregates(i).createAccumulator()
-      i += 1
-    }
+    accumulators = function.createAccumulators()
 
     val iterator = records.iterator()
 
     while (iterator.hasNext) {
       val record = iterator.next()
+      var i = 0
 
       // accumulate
-      i = 0
-      while (i < aggregates.length) {
-        aggregates(i).accumulate(accumulators(i), record.getField(aggInFields(i)(0)))
-        i += 1
-      }
+      function.accumulate(accumulators, record)
 
       // check if this record is the last record
       if (!iterator.hasNext) {
```

Comment:
Couldn't we use `function.setForwardFields()` to forward the grouping keys to the output?